scvid/module/_probabilisticpca.py

```diff
@@ -0,0 +1,175 @@
+    """
+    A PyroModule that serves as the model for the ProbabilisticPCAPyroModule class.
+
+    Args:
+        n_obs: Number of observations.
+        n_vars: Number of input features.
+        n_components: Number of components to model.
+        mean: Mean of the input data.
+        marginalize_z: Marginalize out latent variable z.
+    """
+
+    def __init__(
+        self,
+        n_obs: int,
+        n_vars: int,
+        n_components: int,
+        mean: torch.Tensor,
+        marginalize_z: bool,
+    ):
+        super().__init__(_PROBABILISTIC_PCA_PYRO_MODULE_NAME)
+
+        self.n_obs = n_obs
+        self.n_vars = n_vars
+        self.n_components = n_components
+        self.mean = mean
+        self.marginalize_z = marginalize_z
+
+        # model parameters
+        self.W = PyroParam(lambda: torch.randn((n_components, n_vars)))
+        self.sigma = PyroParam(
+            lambda: torch.tensor(1.0), constraint=constraints.positive
+        )
+
+    def forward(self, x: torch.Tensor):
+        """Forward pass."""
+        with pyro.plate("cells", size=self.n_obs, subsample_size=x.shape[0]):
+            if self.marginalize_z:
+                pyro.sample(
+                    "counts",
+                    dist.LowRankMultivariateNormal(
+                        loc=self.mean,
+                        cov_factor=self.W.T,
+                        cov_diag=self.sigma**2
+                        * torch.ones(self.n_vars, device=x.device),
+                    ),
+                    obs=x,
+                )
+            else:
+                z = pyro.sample(
+                    "z",
+                    dist.Normal(
+                        torch.zeros(self.n_components, device=x.device), 1
+                    ).to_event(1),
+                )
+                pyro.sample(
+                    "counts",
+                    dist.Normal(self.mean + z @ self.W, self.sigma).to_event(1),
+                    obs=x,
+                )
+
+
+class ProbabilisticPCAPyroGuide(PyroModule):
+    """
+    A PyroModule that serves as the guide for the ProbabilisticPCAPyroModule class.
+
+    Args:
+        n_obs: Number of observations.
+        n_vars: Number of input features.
+        n_components: Number of components to model.
+        mean: Mean of the input data.
+        marginalize_z: Marginalize out latent variable z.
+    """
+
+    def __init__(
+        self,
+        n_obs: int,
+        n_vars: int,
+        n_components: int,
+        mean: torch.Tensor,
+        marginalize_z: bool,
+    ):
+        super().__init__(_PROBABILISTIC_PCA_PYRO_MODULE_NAME)
+
+        self.n_obs = n_obs
+        self.n_vars = n_vars
+        self.n_components = n_components
+        self.mean = mean
+        self.marginalize_z = marginalize_z
+
+        # guide parameters
+        if not self.marginalize_z:
+            self.L = PyroParam(lambda: torch.randn((n_vars, n_components)))
+            self.z_scale = PyroParam(
```

Comment:
Just FYI, this guide is not powerful enough to represent the exact solution, see Eq. (6) here:
http://www.cs.columbia.edu/~blei/seminar/2020-representation/readings/TippingBishop1999.pdf

The exact posterior implies a multivariate normal for `z | x` with the data-independent covariance matrix. Perhaps implement that as well and let the user select the guide type (e.g. `diagonal_normal` and `multivariate_normal`)? Since we expect to work with < 200 PCs in practice, a MVN guide for `z`.

Out of curiosity, did you implement the exact guide from Bishop's paper at some point? Was it too slow?

Reply:
Yeah, the exact guide seems to be much slower.